### PR TITLE
Python2 is deprecated for a while

### PR DESCRIPTION
**Pull request by surak**
This makes rpi-source compatible with python3.

### CODE REVIEW

**Comment by surak** [comment]
This integrates #2 on a single pull request

**Comment by fake-name** [comment]
At least [d60fc65](https://github.com/RPi-Distro/rpi-source/pull/10/commits/d60fc65ec49b2071da849fe17d0a378a3c442278) is broken. You're writing to a file opened in text mode with bytes. Either open it with mode `wb`, and encode the bytes before writing, or just drop the encode.

**Comment by fake-name** [comment]
This is at least partially broken:

```
durr@raspberrypi:~/seeed-linux-dtoverlays $ /usr/local/bin/rpi-source -q --tag-update
Traceback (most recent call last):
  File "/usr/local/bin/rpi-source", line 367, in <module>
    update_tag()
  File "/usr/local/bin/rpi-source", line 170, in update_tag
    writef(update_tag_file, ref)
  File "/usr/local/bin/rpi-source", line 112, in writef
    f.write(str.decode('utf-8'))
AttributeError: 'str' object has no attribute 'decode'

```

It appears that `writef()` is called with both strings and bytes.

File: rpi-source [hunk 1]
```diff
@@ -11,9 +11,9 @@ import os
 import struct
 import sys
 import argparse
-import urllib
-import urllib2
-import urlparse
+import urllib.request, urllib.parse, urllib.error
+import urllib.request, urllib.error, urllib.parse
+import urllib.parse
```
Comment (DrBu7cher) [inline rpi-source:16]:
```suggestion
import urllib.request, urllib.parse, urllib.error
```
This should do the same, right?

**Review by DrBu7cher — CHANGES_REQUESTED** [review]
Please change line 170 to
```python
    writef(update_tag_file, ref.encode("utf-8"))
```

**Comment by surak** [comment]
> Please change line 170 to
> 
> ```python
>     writef(update_tag_file, ref.encode("utf-8"))
> ```

done!

**Comment by MedericFourmy** [comment]
I experienced this error:

```
 *** SoC: BCM2711

 *** Arch: 64-bit

 *** Using: /usr/share/doc/raspberrypi-bootloader/changelog.Debian.gz
Traceback (most recent call last):
  File "/usr/local/bin/rpi-source", line 383, in <module>
    kernel = debian_method(debianlog)
  File "/usr/local/bin/rpi-source", line 278, in debian_method
    fw_rev = re.search(r'firmware as of ([0-9a-fA-F]+)', debian_changelog)
  File "/usr/lib/python3.9/re.py", line 201, in search
    return _compile(pattern, flags).search(string)
TypeError: cannot use a string pattern on a bytes-like object
```

The error disappears when changing line 275 to:

`        debian_changelog = f.read().decode('utf-8')`

**Comment by MedericFourmy** [comment]
But then I encounter this error:

```
 *** SoC: BCM2711

 *** Arch: 64-bit

 *** Using: /usr/share/doc/raspberrypi-bootloader/changelog.Debian.gz

 *** Latest firmware revision: 61966732d03de9b71baf561f920e018b54c241ac

 *** Linux source commit: b'6f921e98008589258f97243fb6658d09750f0a2f'

 *** Download kernel source
--2022-08-04 16:21:51--  https://github.com/raspberrypi/linux/archive/b6f921e98008589258f97243fb6658d09750f0a2f.tar.gz
Resolving github.com (github.com)... 140.82.121.4
Connecting to github.com (github.com)|140.82.121.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://codeload.github.com/raspberrypi/linux/tar.gz/b6f921e98008589258f97243fb6658d09750f0a2f [following]
--2022-08-04 16:21:51--  https://codeload.github.com/raspberrypi/linux/tar.gz/b6f921e98008589258f97243fb6658d09750f0a2f
Resolving codeload.github.com (codeload.github.com)... 140.82.121.9
Connecting to codeload.github.com (codeload.github.com)|140.82.121.9|:443... connected.
HTTP request sent, awaiting response... 404 Not Found
2022-08-04 16:21:52 ERROR 404: Not Found.

Traceback (most recent call last):
  File "/usr/local/bin/rpi-source", line 400, in <module>
    sh("wget %s -O %s https://github.com/raspberrypi/linux/archive/%s.tar.gz" % (("-q" if args.quiet else ""), linux_tar, kernel.git_hash))
  File "/usr/local/bin/rpi-source", line 96, in sh
    subprocess.check_call(cmd, shell=True)
  File "/usr/lib/python3.9/subprocess.py", line 373, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command 'wget  -O /home/gepetto/arducam/virtual_video_node/linux-b'6f921e98008589258f97243fb6658d09750f0a2f'.tar.gz https://github.com/raspberrypi/linux/archive/b'6f921e98008589258f97243fb6658d09750f0a2f'.tar.gz' returned non-zero exit status 8
```

Not sure what to do after that. I finally resorted to using the python2 script.

**Comment by surak** [comment]
> I experienced this error:
> 
> ```
>  *** SoC: BCM2711
> 
>  *** Arch: 64-bit
> 
>  *** Using: /usr/share/doc/raspberrypi-bootloader/changelog.Debian.gz
> Traceback (most recent call last):
>   File "/usr/local/bin/rpi-source", line 383, in <module>
>     kernel = debian_method(debianlog)
>   File "/usr/local/bin/rpi-source", line 278, in debian_method
>     fw_rev = re.search(r'firmware as of ([0-9a-fA-F]+)', debian_changelog)
>   File "/usr/lib/python3.9/re.py", line 201, in search
>     return _compile(pattern, flags).search(string)
> TypeError: cannot use a string pattern on a bytes-like object
> ```
> 
> The error disappears when changing line 275 to:
> 
> ` debian_changelog = f.read().decode('utf-8')`

Done!

File: rpi-source [hunk 2]
```diff
@@ -77,11 +80,11 @@ def debug(str):
 
 def info(str):
     if not args.quiet:
-        print("\n *** %s" % str)
+        print(("\n *** %s" % str))
```
Comment (lurch) [inline rpi-source:83]: I'm fairly sure the extra brackets you've added here and on line 87 aren't necessary?

File: rpi-source [hunk 3]
```diff
@@ -238,13 +255,13 @@ def rpi_update_method(uri):
...
 
-    repo_short = urlparse.urlparse(uri).path
+    repo_short = urllib.parse.urlparse(uri).path
 
     repo_api = "https://api.github.com/repos%s" % repo_short
     repo_raw = "https://raw.githubusercontent.com%s" % repo_short
 
-    kernel.git_hash = download("%s/%s/git_hash" % (repo_raw, fw_rev)).strip()
-    kernel.symvers = "%s/%s/Module%s.symvers" % (repo_raw, fw_rev, processor_type_suffix())
+    kernel.git_hash = download("%s/%s/git_hash" % (repo_raw, fw_rev)).strip().decode('UTF-8')
```
Comment (lurch) [inline rpi-source:263]: Should probably use 'utf-8' here rather than 'UTF-8', for consistency with the other `decode` functions in this script?

**Comment by pelwell** [comment]
Fixed by #17.

**Pull request closed without merging**